README.md

```diff
@@ -9,17 +9,33 @@ Currently supports SDK 14.0 ([flipperzero-firmware@0.77.1](https://github.com/fl
 ## Initial setup
 
 1. Install [`rustup`](https://rust-lang.github.io/rustup/) by following the instructions on [`rustup.rs`](https://rustup.rs/).
-2. Use `rustup` to install the `thumbv7em-none-eabihf` target:
+1. Install the nightly build tool-chain to support the[`different-binary-name`](https://doc.rust-lang.org/cargo/reference/unstable.html#different-binary-name) feature:
     ```
-    rustup target add thumbv7em-none-eabihf
+    rustup toolchain install nightly
+    ```
+1. Install [`cargo-generate`](https://github.com/cargo-generate/cargo-generate):
+    ```
+    cargo install cargo-generate
+    ```
+1. Use `rustup` to install the `thumbv7em-none-eabihf` target to the nightly build:
+    ```
+    rustup target add --toolchain nightly thumbv7em-none-eabihf
     ```
 
-## Use `cargo generate` to clone this template.
-
-```
-cargo generate --git https://github.com/flipperzero-rs/flipperzero-template.git --name my-project
-cd my-project
-```
+## Generate the project
+1. Use `cargo generate` to clone this template:
+    ```
+    cargo generate --git https://github.com/flipperzero-rs/flipperzero-template.git --name my-project
+    ```
+    If the `cargo generate`command fails with a segfault, try cloning the template project locally and use a relative path instead:
```

Comment:
Do you have some more details on this issue? Is there a related cargo or cargo-generate issue?

Reply:
Hi @dcoles, unfortunately not. The only output I get is this:
```
cargo -v generate --git https://github.com/flipperzero-rs/flipperzero-template.git --name my-project

[1]    75631 segmentation fault  cargo generate --git  --name my-project
```
I have no idea how to get more information out of the cargo command, adding the `-v` switch does not help and I cannot find any info on debug information. Nor do I find any crash files.

Running on an MacBook Pro M1 in iTerm.

Reply:
You should be able to get the version information with `cargo --version`.

It's hard to tell if issues like this are to do with nightly or something specific to your local environment.

You might be able to get some [more debugging information](https://stackoverflow.com/questions/8370468/debugging-segmentation-faults-on-a-mac) by looking in `~/Library/Logs/DiagnosticReports/` or running cargo under `lldb`:

```
$ lldb --  cargo generate --git  --name my-project
...
(lldb) r
Process 89510 launched
...
(lldb) bt
<backtrace for the thread that crashed>
```

But good to know that there's a workaround for this.

Reply:
I started googling and found a reference on using `lldb` but no luck so far with that. Need to learn the tool first. New to Rust, this is in fact my first experience of the language and its toolchain... 

Output of `cargo --version`
```
cargo 1.69.0 (6e9a83356 2023-04-12)
```

Thanks for pointing me to `~/Library/Logs/DiagnosticReports/` - should have looked harder... :-). In any case, the relevant file(s) do not tell me much, but perhaps you can get more out of it. Here is one file:
```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               cargo-generate [75662]
Path:                  /Users/USER/*/cargo-generate
Identifier:            cargo-generate
Version:               ???
Code Type:             ARM-64 (Native)
Parent Process:        zsh [20342]
Responsible:           iTerm2 [78437]
User ID:               501

Date/Time:             2023-05-20 08:57:39.6055 +0200
OS Version:            macOS 13.0 (22A380)
Report Version:        12
Anonymous UUID:        F7FD002C-0EF3-5A63-73CF-6B7FEAA2645D

Sleep/Wake UUID:       BD0AC9DF-64B0-42D0-9ACE-5D8BDEEEA2D0

Time Awake Since Boot: 4000000 seconds
Time Since Wake:       286 seconds

System Integrity Protection: enabled

Crashed Thread:        0  main  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [75662]

VM Region Info: 0 is not in any region.  Bytes before following region: 4338188288
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      102938000-102f10000    [ 5984K] r-x/r-x SM=COW  ...argo-generate

Thread 0 Crashed:: main Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x1862b5704 _platform_strlen + 4
1   libsystem_c.dylib             	       0x18614b8c0 strdup + 28
2   cargo-generate                	       0x102d92f54 stdalloc__strdup + 16
3   cargo-generate                	       0x102d797b8 server_setup_from_url + 116
4   cargo-generate                	       0x102d7815c git_http_client_send_request + 440
5   cargo-generate                	       0x102d7749c http_stream_read + 200
6   cargo-generate                	       0x102d7bb54 git_smart__recv_cb + 64
7   cargo-generate                	       0x102d7d97c git_smart__store_refs + 184
8   cargo-generate                	       0x102d7bfd0 git_smart__connect + 320
9   cargo-generate                	       0x102d4bc74 git_remote_connect_ext + 304
10  cargo-generate                	       0x102d115d4 clone_into + 424
11  cargo-generate                	       0x102d11104 git__clone + 764
12  cargo-generate                	       0x1029a5054 cargo_generate::git::utils::clone_git_template_into_temp::h5ea10b85512fd76e + 8644
13  cargo-generate                	       0x1029cf024 cargo_generate::generate::hd8f24e124b527017 + 5028
14  cargo-generate                	       0x1029e181c cargo_generate::main::h2eb4a4ad42c1b22f + 4992
15  cargo-generate                	       0x1029df10c std::sys_common::backtrace::__rust_begin_short_backtrace::hd3697d7b14643f38 + 12
16  cargo-generate                	       0x1029dff48 main + 1200
17  dyld                          	       0x185f5fe50 start + 2544


Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000000000   x1: 0x0000000000000000   x2: 0x000000000000034f   x3: 0x00006000030f4314
    x4: 0x0000000000000003   x5: 0x0000000000000310   x6: 0x0000000000000000   x7: 0x0000000000000240
    x8: 0x0000000102d92f44   x9: 0x00006000030f4310  x10: 0x0000b690b0e10000  x11: 0x00006000030f4000
   x12: 0x0000000000000010  x13: 0x00000000001ff800  x14: 0x00000000000007fb  x15: 0x00000000ab5c8832
   x16: 0x00000001862b5700  x17: 0x00000001e679d408  x18: 0x0000000000000000  x19: 0x0000000000000000
   x20: 0x00006000009f1320  x21: 0x0000000000000000  x22: 0x0000000102f70a90  x23: 0x00000001480081f8
   x24: 0x0000000000000001  x25: 0x0000000000000000  x26: 0x0000000000000000  x27: 0x0000000102f70a90
   x28: 0x000000000000003b   fp: 0x000000016d4bdfd0   lr: 0x000000018614b8c0
    sp: 0x000000016d4bdfb0   pc: 0x00000001862b5704 cpsr: 0x20001000
   far: 0x0000000000000000  esr: 0x92000006 (Data Abort) byte read Translation fault

Binary Images:
       0x1862b4000 -        0x1862bbffb libsystem_platform.dylib (*) <78189670-2f47-332c-9a59-a210157aa97b> /usr/lib/system/libsystem_platform.dylib
       0x18614a000 -        0x1861caff3 libsystem_c.dylib (*) <a3869999-0792-3e09-b3fb-5e87ae4639be> /usr/lib/system/libsystem_c.dylib
       0x102938000 -        0x102f0ffff cargo-generate (*) <7d2cb667-795f-39bf-ba35-23e7c0d40dd0> /Users/USER/*/cargo-generate
       0x185f5a000 -        0x185fe44af dyld (*) <1f56ab4e-f398-3a9c-bfba-fbbd17808963> /usr/lib/dyld

External Modification Summary:
  Calls made by other processes targeting this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by all processes on this machine:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0

VM Region Summary:
ReadOnly portion of Libraries: Total=872.6M resident=0K(0%) swapped_out_or_unallocated=872.6M(100%)
Writable regions: Total=667.3M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=667.3M(100%)

                                VIRTUAL   REGION 
REGION TYPE                        SIZE    COUNT (non-coalesced) 
===========                     =======  ======= 
Kernel Alloc Once                   32K        1 
MALLOC                           155.2M       15 
MALLOC guard page                   96K        5 
MALLOC_MEDIUM (reserved)         120.0M        1         reserved VM address space (unallocated)
MALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)
Stack                             8160K        1 
Stack Guard                       56.0M        1 
VM_ALLOCATE                        160K        3 
__AUTH                             307K       58 
__AUTH_CONST                      3542K      142 
__DATA                            1542K      139 
__DATA_CONST                      4756K      146 
__DATA_DIRTY                       361K       58 
__LINKEDIT                       765.9M        4 
__OBJC_CONST                       289K       36 
__OBJC_RO                         65.1M        1 
__OBJC_RW                         1981K        1 
__TEXT                           106.7M      153 
dyld private memory                256K        1 
shared memory                       16K        1 
===========                     =======  ======= 
TOTAL                              1.6G      768 
TOTAL, minus reserved VM space     1.1G      768 



-----------
Full Report
-----------

{"app_name":"cargo-generate","timestamp":"2023-05-20 08:57:40.00 +0200","app_version":"","slice_uuid":"7d2cb667-795f-39bf-ba35-23e7c0d40dd0","build_version":"","platform":1,"share_with_app_devs":0,"is_first_party":1,"bug_type":"309","os_version":"macOS 13.0 (22A380)","roots_installed":0,"incident_id":"C5E4B379-01F8-4232-8A4E-686CAC15B19A","name":"cargo-generate"}
{
  "uptime" : 4000000,
  "procRole" : "Unspecified",
  "version" : 2,
  "userID" : 501,
  "deployVersion" : 210,
  "modelCode" : "MacBookPro18,3",
  "coalitionID" : 411378,
  "osVersion" : {
    "train" : "macOS 13.0",
    "build" : "22A380",
    "releaseType" : "User"
  },
  "captureTime" : "2023-05-20 08:57:39.6055 +0200",
  "incident" : "C5E4B379-01F8-4232-8A4E-686CAC15B19A",
  "pid" : 75662,
  "translated" : false,
  "cpuType" : "ARM-64",
  "roots_installed" : 0,
  "bug_type" : "309",
  "procLaunch" : "2023-05-20 08:57:39.5194 +0200",
  "procStartAbsTime" : 96086342999457,
  "procExitAbsTime" : 96086344984023,
  "procName" : "cargo-generate",
  "procPath" : "\/Users\/USER\/*\/cargo-generate",
  "parentProc" : "zsh",
  "parentPid" : 20342,
  "coalitionName" : "com.googlecode.iterm2",
  "crashReporterKey" : "F7FD002C-0EF3-5A63-73CF-6B7FEAA2645D",
  "responsiblePid" : 78437,
  "responsibleProc" : "iTerm2",
  "wakeTime" : 286,
  "sleepWakeUUID" : "BD0AC9DF-64B0-42D0-9ACE-5D8BDEEEA2D0",
  "sip" : "enabled",
  "vmRegionInfo" : "0 is not in any region.  Bytes before following region: 4338188288\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      102938000-102f10000    [ 5984K] r-x\/r-x SM=COW  ...argo-generate",
  "exception" : {"codes":"0x0000000000000001, 0x0000000000000000","rawCodes":[1,0],"type":"EXC_BAD_ACCESS","signal":"SIGSEGV","subtype":"KERN_INVALID_ADDRESS at 0x0000000000000000"},
  "termination" : {"flags":0,"code":11,"namespace":"SIGNAL","indicator":"Segmentation fault: 11","byProc":"exc handler","byPid":75662},
  "vmregioninfo" : "0 is not in any region.  Bytes before following region: 4338188288\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      102938000-102f10000    [ 5984K] r-x\/r-x SM=COW  ...argo-generate",
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 0,
  "threads" : [{"triggered":true,"id":48126677,"threadState":{"x":[{"value":0},{"value":0},{"value":847},{"value":105553167598356},{"value":3},{"value":784},{"value":0},{"value":576},{"value":4342755140,"symbolLocation":0,"symbol":"stdalloc__strdup"},{"value":105553167598352},{"value":200732559081472},{"value":105553167597568},{"value":16},{"value":2095104},{"value":2043},{"value":2874968114},{"value":6545954560,"symbolLocation":0,"symbol":"_platform_strlen"},{"value":8161711112},{"value":0},{"value":0},{"value":105553126691616},{"value":0},{"value":4344711824,"symbolLocation":0,"symbol":"git__allocator"},{"value":5502960120},{"value":1},{"value":0},{"value":0},{"value":4344711824,"symbolLocation":0,"symbol":"git__allocator"},{"value":59}],"flavor":"ARM_THREAD_STATE64","lr":{"value":6544472256},"cpsr":{"value":536875008},"fp":{"value":6128656336},"sp":{"value":6128656304},"esr":{"value":2449473542,"description":"(Data Abort) byte read Translation fault"},"pc":{"value":6545954564,"matchesCrashFrame":1},"far":{"value":0}},"name":"main","queue":"com.apple.main-thread","frames":[{"imageOffset":5892,"symbol":"_platform_strlen","symbolLocation":4,"imageIndex":0},{"imageOffset":6336,"symbol":"strdup","symbolLocation":28,"imageIndex":1},{"imageOffset":4566868,"symbol":"stdalloc__strdup","symbolLocation":16,"imageIndex":2},{"imageOffset":4462520,"symbol":"server_setup_from_url","symbolLocation":116,"imageIndex":2},{"imageOffset":4456796,"symbol":"git_http_client_send_request","symbolLocation":440,"imageIndex":2},{"imageOffset":4453532,"symbol":"http_stream_read","symbolLocation":200,"imageIndex":2},{"imageOffset":4471636,"symbol":"git_smart__recv_cb","symbolLocation":64,"imageIndex":2},{"imageOffset":4479356,"symbol":"git_smart__store_refs","symbolLocation":184,"imageIndex":2},{"imageOffset":4472784,"symbol":"git_smart__connect","symbolLocation":320,"imageIndex":2},{"imageOffset":4275316,"symbol":"git_remote_connect_ext","symbolLocation":304,"imageIndex":2},{"imageOffset":4036052,"symbol":"clone_into","symbolLocation":424,"imageIndex":2},{"imageOffset":4034820,"symbol":"git__clone","symbolLocation":764,"imageIndex":2},{"imageOffset":446548,"symbol":"cargo_generate::git::utils::clone_git_template_into_temp::h5ea10b85512fd76e","symbolLocation":8644,"imageIndex":2},{"imageOffset":618532,"symbol":"cargo_generate::generate::hd8f24e124b527017","symbolLocation":5028,"imageIndex":2},{"imageOffset":694300,"symbol":"cargo_generate::main::h2eb4a4ad42c1b22f","symbolLocation":4992,"imageIndex":2},{"imageOffset":684300,"symbol":"std::sys_common::backtrace::__rust_begin_short_backtrace::hd3697d7b14643f38","symbolLocation":12,"imageIndex":2},{"imageOffset":687944,"symbol":"main","symbolLocation":1200,"imageIndex":2},{"imageOffset":24144,"symbol":"start","symbolLocation":2544,"imageIndex":3}]}],
  "usedImages" : [
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6545948672,
    "size" : 32764,
    "uuid" : "78189670-2f47-332c-9a59-a210157aa97b",
    "path" : "\/usr\/lib\/system\/libsystem_platform.dylib",
    "name" : "libsystem_platform.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6544465920,
    "size" : 528372,
    "uuid" : "a3869999-0792-3e09-b3fb-5e87ae4639be",
    "path" : "\/usr\/lib\/system\/libsystem_c.dylib",
    "name" : "libsystem_c.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64",
    "base" : 4338188288,
    "size" : 6127616,
    "uuid" : "7d2cb667-795f-39bf-ba35-23e7c0d40dd0",
    "path" : "\/Users\/USER\/*\/cargo-generate",
    "name" : "cargo-generate"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6542434304,
    "size" : 566448,
    "uuid" : "1f56ab4e-f398-3a9c-bfba-fbbd17808963",
    "path" : "\/usr\/lib\/dyld",
    "name" : "dyld"
  }
],
  "sharedCache" : {
  "base" : 6541787136,
  "size" : 3403612160,
  "uuid" : "5667eb2d-1398-3b38-afff-a929572d4b88"
},
  "vmSummary" : "ReadOnly portion of Libraries: Total=872.6M resident=0K(0%) swapped_out_or_unallocated=872.6M(100%)\nWritable regions: Total=667.3M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=667.3M(100%)\n\n                                VIRTUAL   REGION \nREGION TYPE                        SIZE    COUNT (non-coalesced) \n===========                     =======  ======= \nKernel Alloc Once                   32K        1 \nMALLOC                           155.2M       15 \nMALLOC guard page                   96K        5 \nMALLOC_MEDIUM (reserved)         120.0M        1         reserved VM address space (unallocated)\nMALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)\nStack                             8160K        1 \nStack Guard                       56.0M        1 \nVM_ALLOCATE                        160K        3 \n__AUTH                             307K       58 \n__AUTH_CONST                      3542K      142 \n__DATA                            1542K      139 \n__DATA_CONST                      4756K      146 \n__DATA_DIRTY                       361K       58 \n__LINKEDIT                       765.9M        4 \n__OBJC_CONST                       289K       36 \n__OBJC_RO                         65.1M        1 \n__OBJC_RW                         1981K        1 \n__TEXT                           106.7M      153 \ndyld private memory                256K        1 \nshared memory                       16K        1 \n===========                     =======  ======= \nTOTAL                              1.6G      768 \nTOTAL, minus reserved VM space     1.1G      768 \n",
  "legacyInfo" : {
  "threadTriggered" : {
    "name" : "main",
    "queue" : "com.apple.main-thread"
  }
},
  "trialInfo" : {
  "rollouts" : [
    {
      "rolloutId" : "6112d17137f5d11121dcd4e2",
      "factorPackIds" : {

      },
      "deploymentId" : 240000276
    },
    {
      "rolloutId" : "6297d96be2c9387df974efa4",
      "factorPackIds" : {

      },
      "deploymentId" : 240000008
    }
  ],
  "experiments" : [
    {
      "treatmentId" : "c28e4ee6-1b08-4f90-8e05-2809e78310a3",
      "experimentId" : "6317d2003d24842ff850182a",
      "deploymentId" : 400000012
    },
    {
      "treatmentId" : "6dd670af-0633-45e4-ae5f-122ae4df02be",
      "experimentId" : "64406ba83deb637ac8a04419",
      "deploymentId" : 900000005
    }
  ]
}
}
```

Reply:
For what it's worth, output of `cargo generate` under `lldb`:
```
lldb -- cargo -v generate --git https://github.com/flipperzero-rs/flipperzero-template.git --name my-project

(lldb) target create "cargo"
Current executable set to 'cargo' (arm64).
(lldb) settings set -- target.run-args  "-v" "generate" "--git" "https://github.com/flipperzero-rs/flipperzero-template.git" "--name" "my-project"
(lldb) r
Process 75927 launched: '/Users/csoop/.cargo/bin/cargo' (arm64)
Process 75927 stopped
* thread #2, stop reason = exec
    frame #0: 0x0000000101080950 dyld`_dyld_start
dyld`:
->  0x101080950 <+0>:  mov    x0, sp
    0x101080954 <+4>:  and    sp, x0, #0xfffffffffffffff0
    0x101080958 <+8>:  mov    x29, #0x0
    0x10108095c <+12>: mov    x30, #0x0
Target 0: (cargo) stopped.
(lldb) bt
* thread #2, stop reason = exec
  * frame #0: 0x0000000101080950 dyld`_dyld_start
```

Reply:
I'll log a bug report with `cargp` / `cargo generate` and add the issue link here.

Reply:
Found the problem: my `.gitconfig` contains the following lines:
```
[http]
        proxy =
```
The empty proxy setting does not seem to impact how git works and removing the lines resolves the issue.

I guess the issue I had relates to [generate failed when there is a git proxy #664
](https://github.com/cargo-generate/cargo-generate/issues/664) and that we should either remove the workaround, add a link or just ignore the whole thing in the doc, what do you think?

Reply:
My feeling is this might be an issue in [libgit2](https://github.com/libgit2/libgit2). From the backtrace and your `.gitconfig`, I wouldn't be surprised if something assumed that `proxy` always had a non-empty value.

I'll remove the comment from the docs, since it's unlikely most people will trigger this issue.

Reply:
Great, I guess we can close this conversation, but before doing this, is there anywhere I can ask questions on developing for FURI using Rust?

Just tried creating a ViewPort, FuriMessageQueue etc but running the resulting app in Flipper, it fails with an error "Load Failed - found unsatisfied imports".

Honestly speaking, not even sure I use your Rust API right as everything is wrapped in unsafe code to make the compiler happy... Would have hoped for the unsafe c-code to be hidden behind the Rust API, but perhaps this not even possible or even desirable?

Reply:
(Created [this link](https://forum.flipperzero.one/t/furi-app-development-somfy-rts-remote/16561) over at Flipper Zero, but I am not sure how developer focused this forum is...)

Reply:
> Honestly speaking, not even sure I use your Rust API right as everything is wrapped in unsafe code to make the compiler happy... Would have hoped for the unsafe c-code to be hidden behind the Rust API, but perhaps this not even possible or even desirable?

It is indeed possible, but writing those APIs is a lot of (unpaid) work. We are making progress (you can find all the APIs we've wrapped so far in the `flipperzero` crate), but it's going to take time before everything in the Flipper Zero SDK has a Rust equivalent.

Reply:
Rereading the project README I think I am not doing it right and actually rely on unsafe code unnecessarily. I'll try to improve it. From what I understand now, my code is failing because I am relying on the lower level crates instead of only on the top level one.

These are my first steps not only on FlipperZero but also in Rust, so I am missing a lot of obvious stuff about where to look for documentation and currently learning very basic stuff like instantiating classes, eehh, sorry _structs_ 😅 

In any case, I do realise that this is a lot of (unpaid) work and am really appreciative and try to help where I can. Hence this (trivial) PR, but this is at the level that I am at the moment! 😄